### PR TITLE
Use single event system for client events

### LINKED ISSUE #245: Use a single system for events
Instead of registering multiple systems, we can use a single system by using custom functions (similar to what we do for components).
This will make the public API simpler and more efficient.
Bevy did a similar optimization recently: https://github.com/bevyengine/bevy/pull/12936 and we can benefit from the provided `EventRegistry`.

### PR DESCRIPTION
**Pull request by notmd**
Fix  #245

### CODE REVIEW

File: src/network_event/client_event.rs [hunk 1]
```diff
@@ -119,155 +119,219 @@ impl ClientEventAppExt for App {
...
-) {
-    for event in events.read() {
-        let message = DefaultOptions::new()
-            .serialize(&event)
-            .expect("client event should be serializable");
-
-        trace!("sending event `{}`", any::type_name::<T>());
-        client.send(*channel, message);
-    }
+struct ClientEventFns {
```
Comment (Shatur) [inline src/network_event/client_event.rs:172]: I think that this struct will be the same for server events too. I suggest to move it to the `NetworkEvent` module and rename into `NetworkEventFns`.

File: src/network_event/client_event.rs [hunk 2]
```diff
@@ -119,155 +119,219 @@ impl ClientEventAppExt for App {
...
+            send,
+            resend_locally: resend_locally::<T>,
+            receive,
+            reset: reset::<T>,
+        }
     }
 }
 
+type SendFn = fn(&mut World, u8);
+type ReceiveFn = fn(&mut World, u8);
```
Comment (Shatur) [inline src/network_event/client_event.rs:197]: Same with these aliases.

File: src/network_event/client_event.rs [hunk 3]
```diff
@@ -119,155 +119,219 @@ impl ClientEventAppExt for App {
...
+    world.resource_scope(|_world, mut events: Mut<Events<T>>| {
+        let drained_count = events.drain().count();
+        if drained_count > 0 {
+            warn!("Discarded {drained_count} client events due to a disconnect");
+        }
+    })
+}
+
+#[derive(Resource, Default)]
+struct ClientEventRegistry {
```
Comment (Shatur) [inline src/network_event/client_event.rs:264]: Same with the registry: I would move it upper and store both type of events in a single resource, you won't need parallelization here.

File: src/network_event/client_event.rs [hunk 4]
```diff
@@ -119,155 +119,219 @@ impl ClientEventAppExt for App {
...
-        match DefaultOptions::new().deserialize(&message) {
-            Ok(event) => {
-                trace!(
-                    "applying event `{}` from `{client_id:?}`",
-                    any::type_name::<T>()
-                );
-                client_events.send(FromClient { client_id, event });
-            }
-            Err(e) => debug!("unable to deserialize event from {client_id:?}: {e}"),
+fn map_and_send<T: Event + MapEntities + Serialize + Clone>(world: &mut World, channel: u8) {
```
Comment (Shatur) [inline src/network_event/client_event.rs:155]: Could you put `send` first? Just a small personal preference :)

File: src/network_event/client_event.rs [hunk 5]
```diff
@@ -119,155 +119,219 @@ impl ClientEventAppExt for App {
...
+    world.resource_scope(|world, mut client: Mut<RepliconClient>| {
+        let entity_map = world.get_resource::<ServerEntityMap>().unwrap();
+        let events = world.get_resource::<Events<T>>().unwrap();
```
Comment (Shatur) [inline src/network_event/client_event.rs:158]:
```suggestion
        let entity_map = world.resource::<ServerEntityMap>();
        let events = world.resource::<Events<T>>();
```

File: src/network_event/client_event.rs [hunk 6]
```diff
@@ -119,155 +119,219 @@ impl ClientEventAppExt for App {
...
+        }
     }
 }
 
+type SendFn = fn(&mut World, u8);
+type ReceiveFn = fn(&mut World, u8);
+
+fn send<T: Event + Serialize>(world: &mut World, channel_id: u8) {
+    world.resource_scope(|world, mut client: Mut<RepliconClient>| {
+        let events = world.get_resource::<Events<T>>().unwrap();
```
Comment (Shatur) [inline src/network_event/client_event.rs:201]:
```suggestion
        let events = world.resource::<Events<T>>();
```

File: src/network_event/client_event.rs [hunk 7]
```diff
@@ -119,155 +119,219 @@ impl ClientEventAppExt for App {
...
+fn reset_system(world: &mut World) {
+    world.resource_scope(|world, registry: Mut<ClientEventRegistry>| {
+        for event in &registry.events {
+            (event.reset)(world);
+        }
+    });
 }
 
-impl<T> Copy for ClientEventChannel<T> {}
+pub struct ClientNetWorkEventPlugin;
```
Comment (Shatur) [inline src/network_event/client_event.rs:312]: Let's keep names shorter:
```suggestion
pub struct ClientEventPlugin;
```
Also could you move systems under `impl ClientEventPlugin` for consistency with the rest of the codebase?

File: src/network_event/client_event.rs [hunk 8]
```diff
@@ -119,155 +119,219 @@ impl ClientEventAppExt for App {
...
+            receive,
+            reset: reset::<T>,
+        }
     }
 }
 
+type SendFn = fn(&mut World, u8);
+type ReceiveFn = fn(&mut World, u8);
+
+fn send<T: Event + Serialize>(world: &mut World, channel_id: u8) {
```
Comment (Shatur) [inline src/network_event/client_event.rs:199]: I think we probably should make the API nicer by using two functions. First function will read the resource from the registry by ID (like you tried to do before), iterates over all events and passes it into a second function with the correct type.
And let user to override only the second function.

**Review by Shatur — CHANGES_REQUESTED** [review]

**Comment by Shatur** [comment]
I think we probably should make the API nicer by using two functions. First function will read the resource from the registry by ID (like you tried to do before, like Bevy does for updates), iterates over all events and passes each event into a second function with the correct type. And let user to override only the second function.
The same for both sending and receiving.

Also instead of passing the world, we probably should pass a context struct like we do for components. This context will contain the current tick and `TypeRegistry`.

File: src/network_event/client_event.rs (hunk 4 above)
Comment (notmd) [reply src/network_event/client_event.rs]: I moved it below the `send` function

**Comment by notmd** [comment]
> I think we probably should make the API nicer by using two functions. First function will read the resource from the registry by ID (like you tried to do before, like Bevy does for updates), iterates over all events and passes each event into a second function with the correct type. And let user to override only the second function. The same for both sending and receiving.

I'm not sure I fully understand, but doing that we need to store the `ComponentId` (maybe a few times), which seems unnecessary for the internal API. I suppose app can have maximum few hundred events, does this matter?

> I suspect that `ClientEventFns` will be the same as the upcoming `ServerEventFns`. So I would move it upper to the `network_event` module and rename into `NetworkEventFns`.

The `ServerEventFns` have an additional `pop_from_queue` fn, so this is not applicable.

> Also instead of passing the world, we probably should pass a context struct like we do for components. This context will contain the current tick and `TypeRegistry`.

I'm not get this, can you elaborate more?

**Comment by Shatur** [comment]
> I'm not sure I fully understand, but doing that we need to store the ComponentId

I think that you can store it in the struct with functions. The idea behind it is to provide a nice API for users. Functions customization needed only to provide serialization/deserialization for events that doesn't have `Serialize` / `Deserialize`. For example, events with `Box<dyn Reflect>` inside.

> The ServerEventFns have an additional pop_from_queue fn

Maybe we can unite pop_from_queue and receive? I think it will be possible with the mentioned two-functions approach.

> I'm not get this, can you elaborate more?

Sure! Right now we pass `&mut World` into a single function. But let's consider separation into two functions for sending from server:
1. Accepts `&mut World`, `EventContext`, current tick and channel ID. Obtains `Events<T>` from it, iterates over all, passes each `T` into second serialize function and sends the returned bytes over a channel ID to `RepliconServer`.
2. Accepts `T`, `EventContext`, serializes, returns bytes.
Similar for sending and client functions. Here `EventContext` is a struct with necessary context for serialization. I would store `TypeRegistry` and the current tick there.
Take a look at functions for components API, it's similar. So the idea is to give user access only to necessary logic for serialization/deserialization instead of direct `&mut World` (maybe for sending we could use even only `&World`!). And instead of passing `TypeRegistry` and current tick directly, we also create a "context" struct with parameters. For consistency with components API and to avoid making breaking changes if we will need a new to pass more "context" data into functions.

**Comment by Shatur** [comment]
Superseded by #262.

**Pull request closed without merging**